File: 3.longest-substring-without-repeating-characters.java
// @leet imports start
import java.util.*;
import java.math.*;
// @leet imports end

// @leet start
class Solution {
    public int lengthOfLongestSubstring(String s) {
		Integer last = 0;
		Integer ans = 0;
		Map<Character, Integer> map = new HashMap<>();
		for (int i = 0; i < s.length(); i++) {
			Character c = s.charAt(i);
			if (map.containsKey(c) && map.get(c) >= last) {
				last=map.get(c)+1;
			}
			map.put(c,i);
			ans=Math.max(ans, i-last+1);
		}
		return ans;
    }
}
// @leet end

File: 3.longest-substring-without-repeating-characters.py
# @leet start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        hash={}
        last=0
        ans=0
        for i in range(0,len(s)):
            char=s[i]
            if char in hash and hash.get(char)>=last:
                last=hash.get(char)
            hash[char]=i
            ans=max(ans, i-last+1)
        return ans
        
# @leet end
